### ISSUE TITLE
A problem in the PositionalEncoding model code

### ISSUE THREAD

**Issue by kir1to455** [tanjimin/C.Origami#42]
Hi，
Thank you for developing Corigami !
I have encountered some problems when I use corigami to train my data.

![image](https://github.com/tanjimin/C.Origami/assets/87375686/cf5e85d7-7220-4463-8749-069736c5d946)
After the encoder step，here, the transposed matrix is input into attn.
the matrix x  is : Tensor, shape [batch_size, seq_lenth, embedding_dim], not Tensor, shape [seq_lenth, batch_size, embedding_dim] !
![image](https://github.com/tanjimin/C.Origami/assets/87375686/747c10db-2be6-4fe5-9317-b26d03fcb1c0)
if perform this step: x = x + self.pe[:x.size(0)] will return the wrong location information result.

I think the code may have made an error in transponse after the encoder.
![63062fd2a7490cbc735a0d4005c233f](https://github.com/tanjimin/C.Origami/assets/87375686/dc0679b8-9e90-4316-995b-ba91d10f5f81)

Best wishes,
Kirtio

**Comment by tanjimin** [comment]
Hi @kir1to455 Thank you for raising this issue!

The move `feature_forward` function is actually for adjusting the difference between CNN vs Transformer. CNN will have channels/hiddens as the second feature [batch, hidden, length] and NLP models (including transformer) usually have hidden at the end [batch, length, hidden]. We need to swap the last two axis back and forth.

I think you are correct about the PE which is a separate issue. One simple fix could be change the dimension of the pe loaded and let pytorch handle the broad casting. You can try this:

```python
x = x + self.pe[:x.size(1)].transpose(0, 1)
```

I think the decoder compensated for the pe issue, and I'll definitely fix this in the next version! Let me know if you have other questions.

Jimin

**Comment by kir1to455** [comment]
Hi, Jimin @tanjimin 
Thank you for your reply.
Another question I have is why decoder is used in this article instead of transformerdecoder?
In Decoder, you use Dilated Convolution to get more receptive field.
But why not use transformerdecoder? 

Best wishes,
Kirito

**Comment by tanjimin** [comment]
Hi @kir1to455 , the decoder here is actually a dilated 2D conv resnet. It is very different from the typical transformer unless you consider ViT. I named it decoder because it follows the encoder-decoder architecture and decodes genomic features to the final Hi-C map.
